sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -5970,35 +6032,149 @@ pi_result piEnqueueEventsWaitWithBarrier(pi_queue Queue,
-  if (auto Res = TmpWaitList.createAndRetainPiZeEventList(
-          NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine))
-    return Res;
+  // Helper function for appending a barrier to a command list.
+  auto insertBarrierIntoCmdList =
+      [&Queue](pi_command_list_ptr_t CmdList,
+               const _pi_ze_event_list_t &EventWaitList, pi_event &Event) {
+        if (auto Res = createEventAndAssociateQueue(
+                Queue, &Event, PI_COMMAND_TYPE_USER, CmdList))
+          return Res;
+        Event->WaitList = EventWaitList;
+        ZE_CALL(zeCommandListAppendBarrier,
+                (CmdList->first, Event->ZeEvent, EventWaitList.Length,
+                 EventWaitList.ZeEventList));
+        return PI_SUCCESS;
+      };
 
-  // Get a new command list to be used on this call
+  // Indicator for whether batching is allowed. This may be changed later in
+  // this function, but allow it by default.
   bool OkToBatch = true;
-  pi_command_list_ptr_t CommandList{};
-  if (auto Res = Queue->Context->getAvailableCommandList(
-          Queue, CommandList, UseCopyEngine, OkToBatch))
-    return Res;
 
-  ze_event_handle_t ZeEvent = nullptr;
-  auto Res = createEventAndAssociateQueue(Queue, Event, PI_COMMAND_TYPE_USER,
-                                          CommandList);
-  if (Res != PI_SUCCESS)
-    return Res;
-  ZeEvent = (*Event)->ZeEvent;
-  (*Event)->WaitList = TmpWaitList;
+  // If we have a list of events to make the barrier from, then we can create a
+  // barrier on these and use the resulting event as our future barrier.
+  if (NumEventsInWaitList) {
+    // Get an arbitrary command-list in the queue.
+    pi_command_list_ptr_t CmdList;
+    if (auto Res = Queue->Context->getAvailableCommandList(
+            Queue, CmdList,
+            /*UseCopyEngine=*/false, OkToBatch))
+      return Res;
 
-  ZE_CALL(zeCommandListAppendBarrier,
-          (CommandList->first, ZeEvent, (*Event)->WaitList.Length,
-           (*Event)->WaitList.ZeEventList));
+    // Retain the events as they will be owned by the result event.
+    _pi_ze_event_list_t TmpWaitList;
+    if (auto Res = TmpWaitList.createAndRetainPiZeEventList(
+            NumEventsInWaitList, EventWaitList, Queue,
+            CmdList->second.isCopy(Queue)))
+      return Res;
 
-  // Execute command list asynchronously as the event will be used
-  // to track down its completion.
-  return Queue->executeCommandList(CommandList, false, OkToBatch);
+    // Insert the barrier into the command-list and execute.
+    if (auto Res = insertBarrierIntoCmdList(CmdList, TmpWaitList, *Event))
+      return Res;
+    if (auto Res = Queue->executeCommandList(CmdList, false, OkToBatch))
+      return Res;
+
+    // Retain and save the resulting event for future commands.
+    PI_CALL(piEventRetain(*Event));
+    Queue->ActiveBarriers.push_back(*Event);
+    return PI_SUCCESS;
+  }
+
+  // Since there are no events to explicitly create a barrier for, we are
```

Comment:
why can't you just "insertActiveBarriers" here to wait on everything from prior command-lists?

Reply:
The active barriers could be barriers for specific events, so waiting on these for a queue-wide barrier won't be sufficient. Even if we could determine that the active barriers were a queue-wide barrier, there could have been submitted more work between it and this barrier, so waiting on it won't act as a barrier on the later work.

Reply:
I see, thanks. Have you considered an alternative approach to insert an "active" barrier at the end of each command-list (perhaps replacing a fence there that we already use for tracking command-list completion), and then "insertActiveBarriers" here would be sufficient?

Reply:
I did consider it, but I don't think it accounts for cases where new command lists are created later. Additionally, by doing it this way we can have the active barriers finish before more command lists are used, so we can save the barrier.

Reply:
> I don't think it accounts for cases where new command lists are created later

But each command-list would end with a barrier, including every later command-list too. Like I said today we end it with a fence, so replacing that with a barrier may be even better than what we have today.